### PR TITLE
fix: Stop PARTITION BY and UDTFs that fail from terminating the query

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

Stacked on top of https://github.com/confluentinc/ksql/pull/4821 and https://github.com/confluentinc/ksql/pull/4816, (both now merged)

`PARTITION BY` was recently enhanced to support any expression, rather than just column references.  If the expression expression results in an error, e.g. a UDF that throws an exception, this exception was not being handled - resulting in the query being terminated.

UDTFs were recently introduced. If a UDTF throws an exception, or an exception occurs extracting a UDTFs parameters, e.g. if a parameter is a UDF that throws, then the exception was not being handled - resulting in the query being terminated.

All these exceptions are now handled by logging the error to the processing logger and continuing.

### Testing done 

QTT and unit tests added to cover the new exception handling and logging

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/ThrowingUdtf.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2019 Confluent Inc.
```
Comment (agavra) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/ThrowingUdtf.java]: get with the times! (😉 update your file template)

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/ExpressionMetadata.java [hunk 2]
```diff
@@ -62,16 +64,35 @@ public Expression getExpression() {
...
+      final GenericRow row,
+      final Object defaultValue,
+      final ProcessingLogger logger,
+      final Supplier<String> errorMsg
+  ) {
     try {
       return expressionEvaluator.evaluate(getParameters(row));
     } catch (final InvocationTargetException e) {
-      throw new KsqlException(e.getCause().getMessage(), e.getCause());
+      logger.error(RecordProcessingError.recordProcessingError(errorMsg.get(), e.getCause(), row));
```
Comment (agavra) [inline ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/ExpressionMetadata.java]: I think there's an existing configuration about whether to hard-fail on error or log-and-continue. If there is, we should respect that, and if there isn't we should add one so people can control if they want their application skipping events

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/ThrowingUdtf.java (hunk 1 above)
Comment (big-andy-coates) [reply ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/ThrowingUdtf.java]: My template is a macro..... weird...

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/ExpressionMetadata.java (hunk 2 above)
Comment (big-andy-coates) [reply ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/ExpressionMetadata.java]: It's an interesting point... 

However, the changes I've made in this PR close a few holes where we invoke `evaluate` but don't protect it within a try/catch(log) block. 

The existing places that do this try/catch(log) block do NOT honour any streams exception handling config on whether to exist or log-and-continue.

Hence, while I think it's an interesting point, I also think its outside the scope of this change.  A log of our code / UDFs etc throw NPEs and a variety of other exceptions. We'd need to clean all those up before having a query terminate due to bad data...

Comment (big-andy-coates) [reply ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/ExpressionMetadata.java]: Talked offline - @agavra's happy with this.

**Pull request merged**